### ISSUE TITLE
Idea to improve decompression performance for repeated sequences with short period

### ISSUE THREAD

**Issue by mtopolnik** [Cyan4973/lz4#126]
I am testing the performance of the LZ4 library through the JNI binding provided by jpountz. Currently the Java binding uses r123. I have determined a quite bad degradation of performance when the decompressed data consists of a short byte sequence repeated many times (with a period of 1-32 bytes). In reviewing the memory operations used for decompression I noticed that unaligned memory access is applied, which at least my Intel Core i7 does quite slowly when source and target are close together. As a specific example, for a period length 101 decompression reaches 16 GB/s, but for periods of 3, 5, or 7 the speed drops to 1.8 GB/s. A simple stretch of the same byte value works with 5 GB/s.

To alleviate this issue I have written a wild copying routine which does no reading at all in the copy loop for periods up to 8 and for wider periods uses word-aligned writing. For periods of 3, 5, and 7 I get at least 6 GB/s; for periods of 1, 2, 4, 8 I get ~30 GB/s (vs. 5 for the JNI version). For wider periods, say 13, I get 3.6 GB/s vs. 1.8 GB/s native.

I have opened this issue in case there might be some interest in involving this kind of approach on the native LZ4 project, My code was presented at https://github.com/jpountz/lz4-java/issues/72. It is specific to a little-endian architecture, but is trivial to adapt to big-endian (swapping all `>>>` and `<<` ops).

**Comment by Cyan4973** [comment]
Thanks for the suggestion !
I'll have a look into it.
Rgds

**Comment by Cyan4973** [comment]
I've been looking at your suggested improvement at https://github.com/jpountz/lz4-java/issues/72.
They look promising. But it's not trivial to grasp the entire mechanism in a single read.
Consequently, I've got a few questions.

The first is about measurements. I can't guarantee that Adrien's Java version uses the same tricks as the "C" (native) version, nor that they are translated into the same assembler code by the jvm. Nonetheless, it's the version you compare to. So it becomes a bit more difficult to evaluate the potential improvements from your method.
I see that you have also compared to the "JNI" version, which binds the C code. As you know, there is an additional penalty associated with JNI which becomes more and more prevalent as the sample size of data to compress becomes smaller.

So I would suggest to have some measurement on the "C" native version, if possible, to establish a baseline for future improvements comparisons.

If writing "C" code is not to your taste, you may be interested in trying the integrated benchmark utility. It's accessible using `lz4 -b filename` (presuming you have lz4 command line utility available or compiled). Filename can then be any content, including periods as you propose.

I might as well have a look at it, but I don't know when, due to an already dangerously overcrowded planning.

Note also that latest lz4 version, r130, is supposed to be quite a bit faster than r123. On normal data, expect some gains in the +20% range. Not sure if it's applicable to periodic data though.

The second point is a potential risk : your improvements may very well dramatically improve decoding speed for cyclic data, however, in normal circumstances, cyclic data are just a fraction (and a relatively small one) of matches. The risk is that additional logic creates more complex paths. In the end, branches and CPU mispredictions may cost more than any potential gain from the specialized branch. 

This likely does not show up in specialized tests, since only one path is selected.
As a consequence, once you have validated improved results on a specific scenario, it's also important to test them within a "more generic" context. I recommend the _silesia corpus_ or the _calgary_ corpus for such an exercise, mainly to ensure test reproducibility.

The other questions are more technical and patchy, so let's keep them for a follow up.

Regards

**Comment by carlomedas** [comment]
Very small comment on Java JNI impl: one overhead I saw there when looked a couple months ago, was that it's not making use of Java.nio DirectBuffers, so if I'm not wrong at each decompression step you have double overhead:
1) to copy to JNI C
2) to copy result from JNI C back to Java
By leveraging DirectBuffers this can be avoided (although directbuffers are not trivial to be managed then because of other known memory limitations wrt GC) and make Java version become fast almost as C one. You can see an example of this technique in 4mc project (mine).

**Comment by Cyan4973** [comment]
Pretty good point. 4mc project is here : https://github.com/carlomedas/4mc

**Comment by mtopolnik** [comment]
@Cyan4973 

>I can't guarantee that Adrien's Java version uses the same tricks as the "C" (native) version, nor that they are translated into the same assembler code by the jvm. Nonetheless, it's the version you compare to. So it becomes a bit more difficult to evaluate the potential improvements for "C" code using your method.

Even though I have gained enough understanding (by testing out different variants) to expect Java to map quite directly to more-or-less the same assembly as C would, this is still a valid issue which must be controlled for.

>there is an additional penalty associated with JNI which becomes more and more prevalent as the sample size of data to compress becomes smaller.

This is already controlled for because I always use the same 64K block to compress/decompress. Any O(1) JNI overhead is irrelevant against this much data.

>If writing "C" code is not to your taste

To the contrary, I'd love to write C for this---my main issue is setting up the complete build environment. Also, just as you, I'm pressed for time, this being a side track to my main development. I'm trying to leverage LZ4 to achieve record-breaking speed in reloading runtime state of an in-memory store from SSD. Since the state-of-the-art SSDs already provide 1 GB/s sequential reading per core (up to 16 cores in parallel), the bar is relatively high.

>you may be interested in trying the integrated benchmark utility. It's accessible using `lz4 -b filename`

Yes, I can do this easily and I'll compare my results.

>The risk is that additional logic creates more complex paths. In the end, branches and CPU mispredictions may cost more than any potential gain from the specialized branch.

I worried about the same thing :) I have addressed this by putting the common case first: `if (distance > 32)`. So there should be a very cheap check and no jump to reach the common path. For the other cases I am slightly worried about the overhead of additional checks in the case that only a small amount of data is going to be copied. Some more work on the top-level wild copy function, which does the dispatch to specialized functions, could be due.

>once you have validated improved results on a specific scenario, it's also important to test them within a "more generic" context. I recommend the silesia corpus or the calgary corpus for such an exercise, mainly to ensure test reproducibility.

This has been my concern throughout. Even though not being as thorough as the Silesia corpus, I have used Github HTML and concatenated Java .class files. I have not witnessed any performance degradation in those cases. This is not to say that I refuse to use the Silesia corpus :) I guess that will be the easiest part to arrange.

@carlomedas 

>Very small comment on Java JNI impl: one overhead I saw there when looked a couple months ago, was that it's not making use of Java.nio DirectBuffers, so if I'm not wrong at each decompression step you have double overhead.

This may be correct, but the fact that the results of measurement are highly sensitive to the pattern being compressed disproves the hypothesis that the overheads you mention are a significant factor. Since the overhead is fixed in all cases, and since I'm seeing 30 GB/s best-case throughput, this is not my major concern.

**Comment by carlomedas** [comment]
@mtopolnik I definitely agree with your point; moreover I checked latest version of lz4-java is perfectly working with nio DirectBuffers now: if it's direct buffer it avoids the copies, in other cases it falls back to standard overhead... so just make sure you create java.nio.DirectBuffers in your code and to use the API that takes ByteBuffer src and ByteBuffer dest.

**Comment by Cyan4973** [comment]
> I'm trying to leverage LZ4 to achieve record-breaking speed in reloading runtime state of an in-memory store from SSD. Since the state-of-the-art SSDs already provide 1 GB/s sequential reading per core (up to 16 cores in parallel), the bar is relatively high.

That sounds like a great objective.

> I worried about the same thing :) I have addressed this by putting the common case first: if (distance > 32). So there should be a very cheap check and no jump to reach the common path. 

Correct. If we accept the idea than matches at distance > 32 are vastly more common, they will be very predictable, so the cost of the check will be very small for them.
On the other hand, now overlapping matches at distance < 32 start by paying the cost of a branch misprediction. So they need to play catch up. As you suggest, it's likely to require some data length to be worthwhile. Maybe the copy length itself could be part of the branch decision.

Btw, regarding threshold distance 32, how did you come up with this number ? Why not 16 or 64 for example ? Is it a result of heuristic search, or there is some assembly-level explanation for this ?

Regards

**Comment by mtopolnik** [comment]
Yes, it's heuristic and I expect the explanation to be below assembly-level, descending into the specific design of cache lines, store buffers, etc., and what Intel does to fight misalignment of write addresses. 

In the meantime I have some `lz4 -b` results to show. This is the current LZ4:
```
cyclicdata-01.bi :  10000000 ->     39248 ( 0.39%),15230.0 MB/s , 5335.0 MB/s  
cyclicdata-02.bi :  10000000 ->     39251 ( 0.39%),15965.0 MB/s , 5372.3 MB/s  
cyclicdata-03.bi :  10000000 ->     39254 ( 0.39%),15495.0 MB/s , 1881.2 MB/s  
cyclicdata-04.bi :  10000000 ->     39257 ( 0.39%),16530.0 MB/s , 5345.0 MB/s  
cyclicdata-05.bi :  10000000 ->     39260 ( 0.39%),15375.0 MB/s , 1885.0 MB/s  
cyclicdata-06.bi :  10000000 ->     39263 ( 0.39%),15685.0 MB/s , 1888.1 MB/s  
cyclicdata-07.bi :  10000000 ->     39266 ( 0.39%),16295.0 MB/s , 1876.2 MB/s  
cyclicdata-08.bi :  10000000 ->     39269 ( 0.39%),16150.0 MB/s , 5742.1 MB/s  
cyclicdata-09.bi :  10000000 ->     39272 ( 0.39%),15900.0 MB/s , 1902.1 MB/s  
cyclicdata-10.bi :  10000000 ->     39275 ( 0.39%),15285.0 MB/s , 1902.1 MB/s  
cyclicdata-11.bi :  10000000 ->     39278 ( 0.39%),14935.0 MB/s , 1893.1 MB/s  
cyclicdata-12.bi :  10000000 ->     39281 ( 0.39%),15745.0 MB/s , 1895.0 MB/s  
cyclicdata-13.bi :  10000000 ->     39284 ( 0.39%),15860.0 MB/s , 1902.1 MB/s  
cyclicdata-14.bi :  10000000 ->     39287 ( 0.39%),15745.0 MB/s , 1886.2 MB/s  
cyclicdata-15.bi :  10000000 ->     39293 ( 0.39%),15710.0 MB/s , 1887.2 MB/s  
cyclicdata-16.bi :  10000000 ->     39296 ( 0.39%),15100.0 MB/s ,10530.0 MB/s  
cyclicdata-17.bi :  10000000 ->     39299 ( 0.39%),16160.0 MB/s , 3711.3 MB/s  
cyclicdata-18.bi :  10000000 ->     39302 ( 0.39%),15895.0 MB/s , 3763.1 MB/s  
cyclicdata-19.bi :  10000000 ->     39305 ( 0.39%),15940.0 MB/s , 3736.3 MB/s  
cyclicdata-20.bi :  10000000 ->     39308 ( 0.39%),16240.0 MB/s , 3758.1 MB/s  
cyclicdata-21.bi :  10000000 ->     39311 ( 0.39%),15205.0 MB/s , 3736.3 MB/s  
cyclicdata-22.bi :  10000000 ->     39314 ( 0.39%),15555.0 MB/s , 3695.0 MB/s  
cyclicdata-23.bi :  10000000 ->     39317 ( 0.39%),14365.0 MB/s , 3716.3 MB/s  
cyclicdata-24.bi :  10000000 ->     39319 ( 0.39%),15240.0 MB/s ,13825.0 MB/s  
cyclicdata-25.bi :  10000000 ->     39322 ( 0.39%),15390.0 MB/s , 5540.0 MB/s  
cyclicdata-26.bi :  10000000 ->     39325 ( 0.39%),15145.0 MB/s , 5550.0 MB/s  
cyclicdata-27.bi :  10000000 ->     39328 ( 0.39%),16150.0 MB/s , 5592.2 MB/s  
cyclicdata-28.bi :  10000000 ->     39331 ( 0.39%),16495.0 MB/s , 5577.2 MB/s  
cyclicdata-29.bi :  10000000 ->     39334 ( 0.39%),15860.0 MB/s , 5547.2 MB/s  
cyclicdata-30.bi :  10000000 ->     39337 ( 0.39%),16135.0 MB/s , 5572.2 MB/s  
cyclicdata-31.bi :  10000000 ->     39340 ( 0.39%),14875.0 MB/s , 5507.2 MB/s  
cyclicdata.bin   :  10000000 ->     39248 ( 0.39%),15080.0 MB/s , 5320.0 MB/s  
```

And this is my code (yes, I did write some C :) :
```
cyclicdata-01.bi :  10000000 ->     39248 ( 0.39%),16230.0 MB/s ,20655.0 MB/s  
cyclicdata-02.bi :  10000000 ->     39251 ( 0.39%),16240.0 MB/s ,20625.0 MB/s  
cyclicdata-03.bi :  10000000 ->     39254 ( 0.39%),15860.0 MB/s , 7145.0 MB/s  
cyclicdata-04.bi :  10000000 ->     39257 ( 0.39%),15715.0 MB/s ,19775.0 MB/s  
cyclicdata-05.bi :  10000000 ->     39260 ( 0.39%),15590.0 MB/s , 7135.0 MB/s  
cyclicdata-06.bi :  10000000 ->     39263 ( 0.39%),15750.0 MB/s , 7035.0 MB/s  
cyclicdata-07.bi :  10000000 ->     39266 ( 0.39%),15495.0 MB/s , 7100.0 MB/s  
cyclicdata-08.bi :  10000000 ->     39269 ( 0.39%),16150.0 MB/s ,20130.0 MB/s  
```

I still have a bug in my `misaligned` variant so I'm not showing those.

**Comment by mtopolnik** [comment]
>Maybe the copy length itself could be part of the branch decision.

There's another consideration: even in your current code, you must have a branch point for the case where the distance is so close that you can't just copy 8-by-8 bytes (i.e., distances 1-7). So we definitely need this checked and branch into the `narrow` code path. This complicates the design of the branching cascade.

**Comment by Cyan4973** [comment]
These results are very interesting. I really appreciate to visualize these spike of performance at precise distances (1, 2, 4, 8, 16, 24, wait, where is 32 ? what happens after 32 ?). You are right that explanation might be at microcode level, which also means that outcome might be different for other CPU, such as AMD and ARM, and even for different Intel architectures. Quite some testing ahead...

A downside of `lz4 -b` is that you don't have direct control over the starting alignment position. I haven't checked lately, but it's likely to be a malloc'ed buffer, therefore aligned on 16 bytes boundaries. If you want your sample to start at a different position, you may achieve it by adding some fake noise at the beginning of the sample, hence forcing the overlapping sequence to start at your desired position.


Regards

**Comment by mtopolnik** [comment]
I think I'm OK with the current testing because if I use, say, a period of 9, then the destination address is misaligned by 1, even though source is aligned. My testing showed that, whenever source and destination are misaligned by the same offset, the problem doesn't manifest or is much less pronounced. It goes without saying that this is very, very specific to one architecture :) However, I do believe that code which fights hard for aligned writing will win on many different architectures. It may even turn out that other architectures would benefit even beyond the 32 bytes distance.

In the meantime I've fixed my bug and now I have completely functional code. This is the rest of the measurements:
```
cyclicdata-09.bi :  10000000 ->     39272 ( 0.39%),15885.0 MB/s , 4048.0 MB/s  
cyclicdata-10.bi :  10000000 ->     39275 ( 0.39%),15800.0 MB/s , 4053.0 MB/s  
cyclicdata-11.bi :  10000000 ->     39278 ( 0.39%),15635.0 MB/s , 3995.0 MB/s  
cyclicdata-12.bi :  10000000 ->     39281 ( 0.39%),16915.0 MB/s , 4048.0 MB/s  
cyclicdata-13.bi :  10000000 ->     39284 ( 0.39%),15385.0 MB/s , 4028.0 MB/s  
cyclicdata-14.bi :  10000000 ->     39287 ( 0.39%),14765.0 MB/s , 4038.0 MB/s  
cyclicdata-15.bi :  10000000 ->     39293 ( 0.39%),14405.0 MB/s , 4008.0 MB/s  
cyclicdata-16.bi :  10000000 ->     39296 ( 0.39%),15490.0 MB/s ,10795.0 MB/s  
cyclicdata-17.bi :  10000000 ->     39299 ( 0.39%),14960.0 MB/s , 7070.0 MB/s  
cyclicdata-18.bi :  10000000 ->     39302 ( 0.39%),15205.0 MB/s , 7140.0 MB/s  
cyclicdata-19.bi :  10000000 ->     39305 ( 0.39%),15550.0 MB/s , 7281.4 MB/s  
cyclicdata-20.bi :  10000000 ->     39308 ( 0.39%),15890.0 MB/s , 7260.0 MB/s  
cyclicdata-21.bi :  10000000 ->     39311 ( 0.39%),15855.0 MB/s , 7300.0 MB/s  
cyclicdata-22.bi :  10000000 ->     39314 ( 0.39%),14860.0 MB/s , 7000.0 MB/s  
cyclicdata-23.bi :  10000000 ->     39317 ( 0.39%),14895.0 MB/s , 7150.0 MB/s  
cyclicdata-24.bi :  10000000 ->     39319 ( 0.39%),15010.0 MB/s ,13850.0 MB/s  
cyclicdata-25.bi :  10000000 ->     39322 ( 0.39%),14490.0 MB/s , 8235.0 MB/s  
cyclicdata-26.bi :  10000000 ->     39325 ( 0.39%),16005.0 MB/s , 8180.0 MB/s  
cyclicdata-27.bi :  10000000 ->     39328 ( 0.39%),15755.0 MB/s , 8065.0 MB/s  
cyclicdata-28.bi :  10000000 ->     39331 ( 0.39%),16170.0 MB/s , 8160.0 MB/s  
cyclicdata-29.bi :  10000000 ->     39334 ( 0.39%),14735.0 MB/s , 8090.0 MB/s  
cyclicdata-30.bi :  10000000 ->     39337 ( 0.39%),15170.0 MB/s , 8085.0 MB/s  
cyclicdata-31.bi :  10000000 ->     39340 ( 0.39%),15905.0 MB/s , 8255.0 MB/s  
cyclicdata-32.bi :  10000000 ->     39343 ( 0.39%),15690.0 MB/s ,17230.0 MB/s  
```

**Comment by t-mat** [comment]
Quite interesting work, @mtopolnik !

I doesn't have much time to test #127. But here is a small cyclic data generator for the people who want to test it : https://gist.github.com/t-mat/608b86d60171777542bb

**Comment by Cyan4973** [comment]
Thanks @t-mat !

**Comment by mtopolnik** [comment]
One thought... it could be considered to restate the wild copying function's API in terms of `(BYTE[] buf, int srcOff, int destOff, int destEnd)`, i.e., in a pointer + offset format. In that case the ugly pointer->integer conversion would be avoided. I looked into the logs of the failing checks against my PR and see that for ARM this conversion is reported as broken. From the performance standpoint there should be no difference.

**Comment by Cyan4973** [comment]
Sure,
this is internal API, there is no pressure in keeping it identical.

Also, note that the original wildcopy was meant for literal copy, and adapted to work on matches.
Your version seems to be concentrated on matches.
So it is also possible to make both operations more different,

**Comment by mtopolnik** [comment]
@carlomedas Regarding decompression speed with JNI binding, I've done my best but at least with Adrien's JNI version I can't even reach 1 GB/s. I'm testing with a set of files 8 MB in size (`lz4 -b` reports ~2700 MB/s decompression for the same files); I've used a profiler and see the majority of time is spent within the JNI decompress function. Here's the input stream I used:

https://gist.github.com/mtopolnik/c3742d409f4a22ffa792

In short, reading goes like this:

1. Using a `FileChannel` read data from disk into a direct buffer;
2. have the native JNI code decompress the buffer into another direct buffer;
3. transfer the data from direct buffer to the caller's `byte[]`.

That means that no extra copying step is involved. Could it be that the r123 version of LZ4 is all the explanation I need for this lack of success?

**Comment by mtopolnik** [comment]
BTW @carlomedas, about 4mc code... I'm reviewing `LzDecompressor`, trying to see if I can adapt it for my use. You don't actually seem to be benefitting from the direct buffers, at least not on the input side. You copy user-supplied `byte[]` to your internal direct buffer; you could just `wrap` the `byte[]` and hand that over to JNI, let it handle the native memory itself. On the output side you still probably get some benefit from the direct buffer because the caller may supply arbitrary `byte[]` so you need internal storage. With a special contract with the user you may eliminate even that. For example, with a config param the user could promise to always supply buffers of sufficient size to `decompress()`, getting some speed in return---and eradicating direct buffer OOME.

**Comment by mtopolnik** [comment]
@carlomedas Regarding the direct buffer pools, as a user I'd personally be happier with 4mc breaking the public API barrier and calling `DirectByteBuffer.cleaner().clean()`. If the choice is between facing an OOME in production and possibly needing to hold back a JRE upgrade until my lib's private API access is fixed, I gladly choose the latter.

**Comment by carlomedas** [comment]
@mtopolnik I agree with your points, but please keep in consideration that 4mc is implementing Compressor/Decompressor interfaces for hadoop; and I definitely agree with you the interfaces themselves don't look simple and straightforward, too much stateful stuff; so 4mc can surely be used on client standalone code, but main target is to run inside hadoop cluster as M/R jobs. That's why all your good points about contract changes cannot be taken ahead.
I hope for your use cases it can be quickly adapted to become a real zero-copy compress/decompress for JNI binding.

One more thing I wanted to comment on jpountz lz4, is that it brings precompiled native libs .so inside the jars, to make it easier and straightforward for user to run it. So most probably this explains what Yann was telling about code optimisation... the native code, by default, has not been compiled on your target platform.

**Comment by mtopolnik** [comment]
@carlomedas At least the direct buffer on the input side could be dispensed with at no cost. Also the approach to destroy the direct buffers instead of managing the pool is something that could be applied (subject to a judgement call whether accessing stable, but non-public JDK API is acceptable).

**Comment by carlomedas** [comment]
@mtopolnik let's move to 4mc issue if you want to discuss more on this I'm glad; btw I would agree and it was my first try, but after profiling we saw that using the clean you mentioned was not always guaranteed to release thus it was tending to the OOM again after some time of intense usage

**Comment by Cyan4973** [comment]
Should we close this issue, or do you want to keep that thread opened for future discussions on this optimization ?

**Comment by mtopolnik** [comment]
I'd say close it as I have already invested more than my allowance on this, stealing from my main task. I've learned a lot from this experience, especially an the subject of C code portability. Thanks for listening :)
